File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        post = 1
        for i in range(len(nums)):
            res[i] = res[i - 1] * nums[i - 1] if i - 1 >= 0 else 1
        
        for i in range(len(nums) - 2, -1, -1):
            post *= nums[i + 1]
            res[i] *= post

        return res
        

File: Data Structures & Algorithms/three-integer-sum/submission-1.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        // pick one index i first, then do twosum with hashSet for the rest
        Arrays.sort(nums);
        Set<List<Integer>> res = new HashSet<>();
        for (int i = 0; i < nums.length - 2; i ++) {
            if (i > 0 && nums[i] == nums[i] - 1){
                continue;
            }

            int target =  0 - nums[i];
            Set<Integer> seen = new HashSet<>();
            for (int j = i + 1; j < nums.length; j ++){
                if (seen.contains(target - nums[j])) {
                    List<Integer> triplet = List.of(nums[i], target - nums[j], nums[j]);
                    res.add(triplet);
                }
                seen.add(nums[j]);
            }

        }

        return new ArrayList<>(res);

    }
}
